File: fornituras-api/src/main/java/com/numobiz/solutions/fornituras/modules/officers/service/OfficerService.java
package com.numobiz.solutions.fornituras.modules.officers.service;

import com.numobiz.solutions.fornituras.common.audit.AuditWriter;
import com.numobiz.solutions.fornituras.common.crypto.BlindIndexer;
import com.numobiz.solutions.fornituras.common.exception.BadRequestException;
import com.numobiz.solutions.fornituras.common.exception.ConflictException;
import com.numobiz.solutions.fornituras.common.exception.NotFoundException;
import com.numobiz.solutions.fornituras.common.text.CodeNormalizer;
import com.numobiz.solutions.fornituras.modules.catalog.CatalogCodes;
import com.numobiz.solutions.fornituras.modules.catalog.service.CatalogService;
import com.numobiz.solutions.fornituras.modules.officers.dto.OfficerCreateRequest;
import com.numobiz.solutions.fornituras.modules.officers.dto.OfficerDetail;
import com.numobiz.solutions.fornituras.modules.officers.dto.OfficerSummary;
import com.numobiz.solutions.fornituras.modules.officers.entity.Officer;
import com.numobiz.solutions.fornituras.modules.officers.mapper.OfficerMapper;
import com.numobiz.solutions.fornituras.modules.officers.repository.OfficerRepository;
import jakarta.persistence.criteria.Predicate;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import java.util.function.Function;

/**
 * Lógica del padrón de elementos. Centraliza lo sensible: búsqueda sobre datos cifrados (placa en
 * claro + blind index para CURP/RFC), <b>enmascaramiento de PII por rol</b> (el cliente nunca
 * decide la visibilidad) y <b>auditoría</b> del acceso a la ficha completa y de las altas. La PII
 * en claro (nombre/apellidos/CURP/RFC) se cifra en reposo vía el converter de la entidad (ADR 0006).
 *
 * <p>Sexo y tipo de sangre se resuelven contra el catálogo genérico ({@code SEXO}/{@code TIPO_SANGRE},
 * ADR 0007) mediante {@link CatalogService}, no contra tablas propias.
 */
@Service
@Transactional(readOnly = true)
public class OfficerService {

	private static final String ROLE_PII = "ROLE_ADMIN";

	private final OfficerRepository repository;
	private final OfficerMapper mapper;
	private final BlindIndexer blindIndexer;
	private final CatalogService catalogService;
	private final AuditWriter audit;

	public OfficerService(
			OfficerRepository repository,
			OfficerMapper mapper,
			BlindIndexer blindIndexer,
			CatalogService catalogService,
			AuditWriter audit) {
		this.repository = repository;
		this.mapper = mapper;
		this.blindIndexer = blindIndexer;
		this.catalogService = catalogService;
		this.audit = audit;
	}

	public Page<OfficerSummary> findAll(String q, String municipio, Long sexoId, Pageable pageable) {
		Page<Officer> page = repository.findAll(filterBy(q, municipio, sexoId), pageable);
		List<Officer> content = page.getContent();

		Set<Long> catalogIds = new HashSet<>();
		catalogIds.addAll(ids(content, Officer::getSexoId));
		catalogIds.addAll(ids(content, Officer::getTipoSangreId));
		Map<Long, String> names = catalogService.resolveNames(catalogIds);

		return page.map(o -> mapper.toSummary(
				o,
				names.get(o.getSexoId()),
				o.getTipoSangreId() == null ? null : names.get(o.getTipoSangreId())));
	}

	/** Devuelve la ficha (enmascarada por rol) y <b>audita</b> el acceso (FR-006). */
	public OfficerDetail findById(Long id) {
		Officer officer = getOrThrow(id);
		audit.record("VIEW_OFFICER", id);
		return toDetail(officer);
	}

	@Transactional
	public OfficerDetail create(OfficerCreateRequest request) {
		String placaNormalizada = CodeNormalizer.normalize(request.placa());
		if (placaNormalizada.isBlank()) {
			throw new BadRequestException("La placa no es válida.");
		}
		if (repository.existsByPlacaNormalizada(placaNormalizada)) {
			throw new ConflictException("Ya existe un elemento con la placa: " + request.placa());
		}
		String curpIdx = blindIndexer.index(request.curp());
		if (curpIdx != null && repository.existsByCurpIdx(curpIdx)) {
			throw new ConflictException("Ya existe un elemento con esa CURP.");
		}
		String rfcIdx = blindIndexer.index(request.rfc());
		if (rfcIdx != null && repository.existsByRfcIdx(rfcIdx)) {
			throw new ConflictException("Ya existe un elemento con ese RFC.");
		}
		validateCatalogs(request.sexoId(), request.tipoSangreId());

		Officer officer = new Officer();
		officer.setNombre(request.nombre().trim());
		officer.setApellidoPaterno(request.apellidoPaterno().trim());
		officer.setApellidoMaterno(blankToNull(request.apellidoMaterno()));
		officer.setPlaca(request.placa().trim().toUpperCase(Locale.ROOT));
		officer.setPlacaNormalizada(placaNormalizada);
		officer.setCurp(upperOrNull(request.curp()));
		officer.setCurpIdx(curpIdx);
		officer.setRfc(upperOrNull(request.rfc()));
		officer.setRfcIdx(rfcIdx);
		officer.setSexoId(request.sexoId());
		officer.setTipoSangreId(request.tipoSangreId());
		officer.setMunicipio(blankToNull(request.municipio()));
		officer.setEstado(blankToNull(request.estado()));
		officer.setFotoUrl(blankToNull(request.fotoUrl()));
		officer.setActive(true);

		Officer saved = repository.save(officer);
		audit.record("CREATE_OFFICER", saved.getId());
		return toDetail(saved);
	}

	/** El sexo y el tipo de sangre deben ser valores activos de sus catálogos (SEXO/TIPO_SANGRE). */
	private void validateCatalogs(Long sexoId, Long tipoSangreId) {
		catalogService.requireActiveItem(sexoId, CatalogCodes.SEXO);
		if (tipoSangreId != null) {
			catalogService.requireActiveItem(tipoSangreId, CatalogCodes.TIPO_SANGRE);
		}
	}

	private OfficerDetail toDetail(Officer officer) {
		String sexoNombre = catalogService.resolveName(officer.getSexoId());
		String tipoSangreEtiqueta = catalogService.resolveName(officer.getTipoSangreId());
		return mapper.toDetail(officer, sexoNombre, tipoSangreEtiqueta, canViewPii());
	}

	/** El servidor decide la visibilidad de la PII a partir del rol (solo ADMIN ve CURP/RFC). */
	private boolean canViewPii() {
		Authentication auth = SecurityContextHolder.getContext().getAuthentication();
		return auth != null && auth.getAuthorities().stream()
				.anyMatch(authority -> ROLE_PII.equals(authority.getAuthority()));
	}

	private Officer getOrThrow(Long id) {
		return repository.findById(id)
				.orElseThrow(() -> new NotFoundException("Elemento no encontrado: " + id));
	}

	private Specification<Officer> filterBy(String q, String municipio, Long sexoId) {
		return (root, query, cb) -> {
			List<Predicate> predicates = new ArrayList<>();
			if (municipio != null && !municipio.isBlank()) {
				predicates.add(cb.like(cb.upper(root.get("municipio")),
						"%" + municipio.trim().toUpperCase(Locale.ROOT) + "%"));
			}
			if (sexoId != null) {
				predicates.add(cb.equal(root.get("sexoId"), sexoId));
			}
			if (q != null && !q.isBlank()) {
				List<Predicate> matches = new ArrayList<>();
				matches.add(cb.like(root.get("placaNormalizada"), "%" + CodeNormalizer.normalize(q) + "%"));
				String idx = blindIndexer.index(q);
				if (idx != null) {
					matches.add(cb.equal(root.get("curpIdx"), idx));
					matches.add(cb.equal(root.get("rfcIdx"), idx));
				}
				predicates.add(cb.or(matches.toArray(new Predicate[0])));
			}
			return cb.and(predicates.toArray(new Predicate[0]));
		};
	}

	private String blankToNull(String value) {
		return (value == null || value.isBlank()) ? null : value.trim();
	}

	private String upperOrNull(String value) {
		return (value == null || value.isBlank()) ? null : value.trim().toUpperCase(Locale.ROOT);
	}

	private Set<Long> ids(Collection<Officer> content, Function<Officer, Long> extractor) {
		Set<Long> result = new HashSet<>();
		for (Officer o : content) {
			Long id = extractor.apply(o);
			if (id != null) {
				result.add(id);
			}
		}
		return result;
	}
}

File: fornituras-api-dotnet/src/Fornituras.Api/Services/OfficerService.cs
using Fornituras.Api.Common;
using Fornituras.Api.Common.Crypto;
using Fornituras.Api.Common.Text;
using Fornituras.Api.Data;
using Fornituras.Api.Data.Entities;
using Fornituras.Api.Dto;
using Fornituras.Api.Security;
using Microsoft.EntityFrameworkCore;

namespace Fornituras.Api.Services;

public sealed class OfficerService(
    ApplicationDbContext db,
    CatalogService catalogService,
    BlindIndexer blindIndexer,
    CurrentUserService currentUser,
    IAuditWriter audit) : IOfficerService
{
    public Task<PageResult<OfficerSummary>> FindAllAsync(
        string? q,
        string? municipio,
        long? sexoId,
        PaginationQuery pagination,
        CancellationToken cancellationToken = default) =>
        FindAllInternalAsync(q, municipio, sexoId, pagination.Page, pagination.Size, cancellationToken);

    private async Task<PageResult<OfficerSummary>> FindAllInternalAsync(
        string? q,
        string? municipio,
        long? sexoId,
        int page,
        int size,
        CancellationToken cancellationToken = default)
    {
        var query = db.Officers.AsNoTracking().Where(o => o.Active);

        if (!string.IsNullOrWhiteSpace(q))
        {
            var normalizedPlaca = CodeNormalizer.Normalize(q);
            var curpIdx = blindIndexer.Index(q);
            var rfcIdx = blindIndexer.Index(q);

            query = query.Where(o =>
                o.PlacaNormalizada.Contains(normalizedPlaca) ||
                (curpIdx != null && o.CurpIdx == curpIdx) ||
                (rfcIdx != null && o.RfcIdx == rfcIdx));
        }

        if (!string.IsNullOrWhiteSpace(municipio))
        {
            var upper = municipio.Trim().ToUpperInvariant();
            query = query.Where(o => o.Municipio != null && o.Municipio.ToUpper().Contains(upper));
        }

        if (sexoId.HasValue)
        {
            query = query.Where(o => o.SexoId == sexoId.Value);
        }

        query = query.OrderBy(o => o.Placa);
        var total = await query.CountAsync(cancellationToken);
        var officers = await query.Skip(page * size).Take(size).ToListAsync(cancellationToken);
        var summaries = await MapSummariesAsync(officers, cancellationToken);
        return PageResult<OfficerSummary>.From(summaries, total, page, size);
    }

    public async Task<OfficerDetail> FindByIdAsync(long id, CancellationToken cancellationToken = default)
    {
        var officer = await db.Officers.AsNoTracking()
            .FirstOrDefaultAsync(o => o.Id == id, cancellationToken)
            ?? throw new NotFoundException($"Officer not found: {id}");

        await audit.RecordAsync("VIEW_OFFICER", officer.Id, cancellationToken);
        return await MapDetailAsync(officer, cancellationToken);
    }

    public async Task<OfficerDetail> CreateAsync(
        OfficerCreateRequest request,
        CancellationToken cancellationToken = default)
    {
        var placaNorm = CodeNormalizer.Normalize(request.Placa);
        if (string.IsNullOrEmpty(placaNorm))
        {
            throw new BadRequestException("La placa es obligatoria.");
        }

        if (await db.Officers.AnyAsync(o => o.PlacaNormalizada == placaNorm, cancellationToken))
        {
            throw new ConflictException($"Ya existe un elemento con la placa '{request.Placa.Trim()}'.");
        }

        await catalogService.RequireActiveItemAsync(request.SexoId, CatalogCodes.Sexo, cancellationToken);
        if (request.TipoSangreId.HasValue)
        {
            await catalogService.RequireActiveItemAsync(
                request.TipoSangreId.Value, CatalogCodes.TipoSangre, cancellationToken);
        }

        var curpIdx = blindIndexer.Index(request.Curp);
        if (curpIdx is not null &&
            await db.Officers.AnyAsync(o => o.CurpIdx == curpIdx, cancellationToken))
        {
            throw new ConflictException("Ya existe un elemento con esa CURP.");
        }

        var rfcIdx = blindIndexer.Index(request.Rfc);
        if (rfcIdx is not null &&
            await db.Officers.AnyAsync(o => o.RfcIdx == rfcIdx, cancellationToken))
        {
            throw new ConflictException("Ya existe un elemento con ese RFC.");
        }

        var now = DateTime.UtcNow;
        var officer = new Officer
        {
            Nombre = PiiCipher.Encrypt(request.Nombre.Trim())!,
            ApellidoPaterno = PiiCipher.Encrypt(request.ApellidoPaterno.Trim())!,
            ApellidoMaterno = PiiCipher.Encrypt(request.ApellidoMaterno?.Trim()),
            Placa = request.Placa.Trim(),
            PlacaNormalizada = placaNorm,
            Curp = PiiCipher.Encrypt(request.Curp?.Trim()),
            CurpIdx = curpIdx,
            Rfc = PiiCipher.Encrypt(request.Rfc?.Trim()),
            RfcIdx = rfcIdx,
            SexoId = request.SexoId,
            TipoSangreId = request.TipoSangreId,
            Municipio = request.Municipio?.Trim(),
            Estado = request.Estado?.Trim(),
            FotoUrl = request.FotoUrl?.Trim(),
            Active = true,
            CreatedAt = now,
            UpdatedAt = now
        };

        db.Officers.Add(officer);
        await db.SaveChangesAsync(cancellationToken);
        await audit.RecordAsync("CREATE_OFFICER", officer.Id, cancellationToken);
        return await MapDetailAsync(officer, cancellationToken);
    }

    private bool CanViewPii() => RolePolicy.CanViewFullPii(currentUser.User);

    private async Task<IReadOnlyList<OfficerSummary>> MapSummariesAsync(
        IReadOnlyList<Officer> officers,
        CancellationToken cancellationToken)
    {
        var sexoIds = officers.Select(o => o.SexoId).Distinct();
        var sangreIds = officers.Where(o => o.TipoSangreId.HasValue).Select(o => o.TipoSangreId!.Value).Distinct();
        var sexoNames = await catalogService.ResolveNamesAsync(sexoIds, cancellationToken);
        var sangreNames = await catalogService.ResolveNamesAsync(sangreIds, cancellationToken);

        return officers.Select(o =>
        {
            var nombre = DecryptName(o);
            return new OfficerSummary(
                o.Id,
                nombre,
                o.Placa,
                sexoNames.GetValueOrDefault(o.SexoId),
                o.TipoSangreId.HasValue ? sangreNames.GetValueOrDefault(o.TipoSangreId.Value) : null,
                o.Municipio,
                o.Estado,
                o.FotoUrl,
                o.Active);
        }).ToList();
    }

    private async Task<OfficerDetail> MapDetailAsync(Officer o, CancellationToken cancellationToken)
    {
        var unmask = CanViewPii();
        var nombre = PiiCipher.Decrypt(o.Nombre) ?? string.Empty;
        var apPat = PiiCipher.Decrypt(o.ApellidoPaterno) ?? string.Empty;
        var apMat = PiiCipher.Decrypt(o.ApellidoMaterno);
        var curp = PiiCipher.Decrypt(o.Curp);
        var rfc = PiiCipher.Decrypt(o.Rfc);

        if (!unmask)
        {
            curp = PiiMasker.Mask(curp);
            rfc = PiiMasker.Mask(rfc);
        }

        var sexoNombre = await catalogService.ResolveNameAsync(o.SexoId, cancellationToken);
        string? sangre = o.TipoSangreId.HasValue
            ? await catalogService.ResolveNameAsync(o.TipoSangreId.Value, cancellationToken)
            : null;

        var nombreCompleto = string.Join(" ",
            new[] { nombre, apPat, apMat }.Where(s => !string.IsNullOrWhiteSpace(s)));

        return new OfficerDetail(
            o.Id,
            nombre,
            apPat,
            apMat,
            nombreCompleto,
            o.Placa,
            o.SexoId,
            sexoNombre,
            o.TipoSangreId,
            sangre,
            o.Municipio,
            o.Estado,
            curp,
            rfc,
            !unmask,
            o.FotoUrl,
            o.Active,
            o.CreatedAt,
            o.UpdatedAt);
    }

    private static string DecryptName(Officer o)
    {
        var parts = new[]
        {
            PiiCipher.Decrypt(o.Nombre),
            PiiCipher.Decrypt(o.ApellidoPaterno),
            PiiCipher.Decrypt(o.ApellidoMaterno)
        }.Where(s => !string.IsNullOrWhiteSpace(s));
        return string.Join(" ", parts);
    }
}
